accounts/accounts/controllers/registration.py

```diff
@@ -0,0 +1,303 @@
+            msg = 'Registration failed'
+            raise InternalServerError(msg) from e  # type: ignore
+
+        # Log the user in.
+        session, cookie = _login(user, auth, ip)
+        c_session, c_cookie = _login_classic(user, auth, ip)
+        data.update({
+            'cookies': {
+                'session_cookie': (cookie, session.expires),
+                'classic_cookie': (c_cookie, c_session.expires)
+            },
+            'user_id': user.user_id
+        })
+        return data, status.HTTP_201_CREATED, {}
+    return data, status.HTTP_200_OK, {}
+
+
+def view_profile(user_id: str, session: domain.Session) -> ResponseData:
+    """Handle requests to view a user's profile."""
+    user = users.get_user_by_id(user_id)
+    return {'user': user}, status.HTTP_200_OK, {}
+
+
+def edit_profile(method: str, user_id: str, session: domain.Session,
+                 params: Optional[MultiDict] = None,
+                 ip: Optional[str] = None) -> ResponseData:
+    """Handle requests to update a user's profile."""
+    if method == 'GET':
+        user = users.get_user_by_id(user_id)
+        form = ProfileForm.from_domain(user)
+        data = {'form': form, 'user_id': user_id}
+    elif method == 'POST':
+        form = ProfileForm(params)
+        data = {'form': form, 'user_id': user_id}
+
+        if not form.validate():
+            return data, status.HTTP_400_BAD_REQUEST, {}
+
+        if form.user_id.data != user_id:
+            msg = 'User ID in request does not match'
+            raise BadRequest(msg)  # type: ignore
+
+        user = form.to_domain()
+        try:
+            user, auth = users.update(user)
+        except Exception as e:
+            data['error'] = 'Could not save user profile; please try again'
+            return data, status.HTTP_500_INTERNAL_SERVER_ERROR, {}
+
+        # We need a new session, to update user's data.
+        _logout(session.session_id)
+        new_session, new_cookie = _login(user, auth, ip)
+        data.update({'cookies': {
+            'session_cookie': (new_cookie, new_session.expires)
+        }})
+        return data, status.HTTP_303_SEE_OTHER, {}
+    return data, status.HTTP_200_OK, {}
+
+
+class ProfileForm(Form):
+    """User registration form."""
+
+    COUNTRIES = [('', '')] + \
+        [(country.alpha_2, country.name) for country in pycountry.countries]
+    RANKS = [('', '')] + domain.UserProfile.RANKS
+    GROUPS = [
+        (key, group['name']) for key, group in taxonomy.GROUPS.items()
+        if not group.get('is_test', False)
+    ]
+    CATEGORIES = [
+        (archive['name'], [
+            (category_id, category['name'])
+            for category_id, category in taxonomy.CATEGORIES_ACTIVE.items()
+            if category['in_archive'] == archive_id
+        ])
+        for archive_id, archive in taxonomy.ARCHIVES_ACTIVE.items()
+    ]
+    """Categories grouped by archive."""
+
+    user_id = HiddenField('User ID')
+
+    email = StringField(
```

Comment:
The classic system does some additional email address validation; it blocks some (like the example.org domain) and pushes back on others (like free email providers). 

<img width="992" alt="screenshot 2018-07-09 10 44 13" src="https://user-images.githubusercontent.com/746253/42457286-1a7db0b2-8365-11e8-8644-0c0f53c244aa.png">

<img width="1001" alt="screenshot 2018-07-09 10 44 31" src="https://user-images.githubusercontent.com/746253/42457295-22063f2a-8365-11e8-8842-5b73ac38a760.png">

Additionally, before their account is active, users must verify their email address by clicking on a link sent to that address.

Reply:
I don't feel that strongly about carrying the "do you have another email address" check forward since it represents another barrier, but the admin team may have a different opinion.

Reply:
Created https://culibrary.atlassian.net/browse/ARXIVNG-974